virt-api: replace hardcoded patch operations with patchset

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Use the patch package to replace the hardcoded patch operations. This improves the readability and make the code more cohesive.

#### Before this PR:

Hardcoded operation in virt-api package.

#### After this PR:
Replaced all the patch operations with the `patch.PatchSet` in the virt-api package.

Fixes #
Partially address: https://github.com/kubevirt/kubevirt/issues/11654

**N.B.**
The vmi mutation webhook now does not return the patch
field if no patch are requested. Previously, it always
returns a patch, eventually "null".

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

/cc @alicefr @0xFelix

alicefr commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -212,60 +212,31 @@ func (app *SubresourceAPIApp) getVirtHandlerConnForVMI(vmi *v1.VirtualMachineIns
 }
 
 func getChangeRequestJson(vm *v1.VirtualMachine, changes ...v1.VirtualMachineStateChangeRequest) (string, error) {
```
Would it make sense to return directly byte here instead of string?

fossedihelm replied:
Done! thanks

alicefr commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -933,15 +935,15 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 					},
 				},
-				"[{ \"op\": \"test\", \"path\": \"/spec/volumes\", \"value\": [{\"name\":\"existingvol\",\"persistentVolumeClaim\":{\"claimName\":\"testpvcdiskclaim\"}}]}, { \"op\": \"test\", \"path\": \"/spec/domain/devices/disks\", \"value\": [{\"name\":\"existingvol\"}]}, { \"op\": \"replace\", \"path\": \"/spec/volumes\", \"value\": [{\"name\":\"existingvol\",\"persistentVolumeClaim\":{\"claimName\":\"te ...
+				"[{\"op\":\"test\",\"path\":\"/spec/volumes\",\"value\":[{\"name\":\"existingvol\",\"persistentVolumeClaim\":{\"claimName\":\"testpvcdiskclaim\"}}]},{\"op\":\"test\",\"path\":\"/spec/domain/devices/disks\",\"value\":[{\"name\":\"existingvol\"}]},{\"op\":\"replace\",\"path\":\"/spec/volumes\",\"value\":[{\"name\":\"existingvol\",\"persistentVolumeClaim\":{\"claimName\":\"testpvcdiskclaim\"}},{ ...
```
I would personally use the patchset type also in this case, because we can avoid this formatting and ordering of the type

fossedihelm replied:
I tried to give it a try, but the result is that each `Entry` became incredibly huge and the readability is seriously affected. I would leave it as is for the moment. Maybe we can try to address it in a separate PR. WDYT?

alicefr replied:
I'm afraid that we'll forget about it. Any reasons to not handle it in this PR? i think you can define some functions that hides the values assignment. For example:
```golang

type vol struct {
    name string
    pvc string
}

func testAndReplaceVols(oldVols, newVols []vol) []byte{
     oldVol []v1.Volume
     oldDisk []v1.Disk
     newVol []v1.Volume
     newDisk []v1.Disk
     .... create the 2 volumes set and the patch
}
```
then in the entry you have:
```golang
testAndReplaceVols([]vol{{"existing","testpvcdiskclaim }}, ([]vol{{"existing","testpvcdiskclaim }, {"vol1"}}
```

fossedihelm replied:
Sure, will do soon. Thanks

fossedihelm replied:
From a second look, it appears that my worries were unfounded.
I opted for the patchSet without using the `testAndReplaceVols()`, because it is only two usage for the volumes.

alicefr replied:
I personally like more this version with the struct even if the Entry is larger, thanks!

fossedihelm replied:
Yeah I agree, I was scared previously if there were a high number of those larger `Entry`.
Thank you!

alicefr commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go:
```diff
@@ -129,7 +129,7 @@ var _ = Describe("Pod eviction admitter", func() {
 		Expect(kubeClient.Fake.Actions()).To(HaveLen(1))
 
-		expectedJSONPatchData := fmt.Sprintf(`[{ "op": "add", "path": "/status/evacuationNodeName", "value": "%s" }]`, vmi.Status.NodeName)
+		expectedJSONPatchData := fmt.Sprintf(`[{"op":"add","path":"/status/evacuationNodeName","value":"%s"}]`, vmi.Status.NodeName)
```
@fossedihelm I think this should be easier to replace with the patchset

alicefr commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -1390,7 +1387,7 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 					Phase:     v1.MemoryDumpCompleted,
 				},
-				"[{ \"op\": \"test\", \"path\": \"/status/memoryDumpRequest\", \"value\": {\"claimName\":\"vol1\",\"phase\":\"Completed\"}}, { \"op\": \"replace\", \"path\": \"/status/memoryDumpRequest\", \"value\": {\"claimName\":\"vol1\",\"phase\":\"Associating\"}}]",
+				"[{\"op\":\"test\",\"path\":\"/status/memoryDumpRequest\",\"value\":{\"claimName\":\"vol1\",\"phase\":\"Completed\"}},{\"op\":\"replace\",\"path\":\"/status/memoryDumpRequest\",\"value\":{\"claimName\":\"vol1\",\"phase\":\"Associating\"}}]",
```
Same suggestion here to create a function that creates the assignement and the patch. In this case something like:
```golang
type memoryDumpRequest struct {
       claim string
       phase string
}


func testAndReplace(oldReq, newReq memoryDumpRequest)
```
Then the entry becomes:
```golag
testAndReplace(memoryDumpRequest{"vol1", "Completed"}, memoryDumRequest("vol1", "Associating")
```

0xFelix commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -974,28 +944,18 @@ func generateVMVolumeRequestPatch(vm *v1.VirtualMachine, volumeRequest *v1.Virtu
-	return verb
+	patchBytes, err := patchSet.GeneratePayload()
+
+	return string(patchBytes), err
```
Makes sense to change this func to return bytes too?

fossedihelm replied:
Yes, done! Thanks

0xFelix commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -211,61 +211,31 @@ func (app *SubresourceAPIApp) getVirtHandlerConnForVMI(vmi *v1.VirtualMachineIns
-
-	verb := "add"
+func getChangeRequestJson(vm *v1.VirtualMachine, changes ...v1.VirtualMachineStateChangeRequest) ([]byte, error) {
+	var patchSet patch.PatchSet
```
Shouldn't this be using `patch.New()`?

fossedihelm replied:
Nice catch!

fossedihelm commented:
@alicefr @0xFelix I addressed your comments (and more).
If I haven't missed anything, now there aren't any hardcode operations in the whole virt-api pkg (test file included).
Actually, there is only one; but it that case the test purpose is to check an invalid patch request, and the patchSet does not allow us to create wrong patch (obviously).
For reference this is the test: 
```
It("should reject an invalid patch", func() {
	const invalidPatch = `{"op": "remove", "path": "/spec/running" : "illegal-field"}`
	restore.Spec.Patches = []string{invalidPatch}

	ar := createRestoreAdmissionReview(restore)
	resp := createTestVMRestoreAdmitter(config, vm, snapshot).Admit(ar)
	Expect(resp.Allowed).To(BeFalse())
	Expect(resp.Result.Details.Causes).To(HaveLen(1))
	Expect(resp.Result.Details.Causes[0].Field).To(Equal("spec.patches"))
})
```
PTAL. Thank you!

0xFelix commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -953,14 +981,17 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 				Expect(err).To(HaveOccurred())
-			} else {
-				Expect(err).ToNot(HaveOccurred())
+				Expect(generatedPatch).To(Equal([]byte{}))
```
```suggestion
				Expect(generatedPatch).To(BeEmpty())
```

I have a feeling like checking for zero length is cleaner and less error prone?

0xFelix commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -953,14 +981,17 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 			}
 
-			Expect(patch).To(Equal(expectedPatch))
+			Expect(err).ToNot(HaveOccurred())
+			expectedPatch, err := expectedPatchSet.GeneratePayload()
+			Expect(err).To(Not(HaveOccurred()))
```
nit: ToNot() or To(Not())?

fossedihelm replied:
"ToNot() or To(Not()) that is the question" [cit. Hamlet]

0xFelix replied:
Or NotTo()?

0xFelix commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -1368,18 +1437,28 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 				Expect(err).To(HaveOccurred())
-			} else {
-				Expect(err).ToNot(HaveOccurred())
+				Expect(patch).To(Equal([]byte{}))
```
Same

0xFelix commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -1368,18 +1437,28 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 			}
-			Expect(patch).To(Equal(expectedPatch))
+
+			Expect(err).ToNot(HaveOccurred())
+			patchBytes, err := expectedPatchSet.GeneratePayload()
+			Expect(err).To(Not(HaveOccurred()))
```
Same

fossedihelm commented:
Thank you @0xFelix, PTAL

alicefr commented:
Looks good to me
/approve

0xFelix reviewed (approved):
/lgtm

Thanks!
